### DISCUSSION TITLE
Order of items in config.json?

### DISCUSSION THREAD

**Discussion by kubedzero** [Pryaxis/TShock#1964] [General]
I wanted to get some background on the design of the config.json. Since it doesn't ship with the current version of Tshock and instead gets auto-generated at server launch time, we aren't able to see the contents until after the server is launched for the first time. That also means the server has write permissions rather than just read permissions, even though it's not really necessary from what I've seen. The server never updates the JSON based on user-input settings in the console, does it?

Specifically regarding the title of this topic, I was bothered by the seemingly random order of the items in the config.json and used a JSON sorting tool to order them alphabetically. I found it easier to locate and manage the different settings I wanted to edit that way. However, upon server restart (or even just the `/reload` command) the JSON went back to its previous order.

1. How is the current `config.json` sorted?
1. Has any thought been given to shipping a `config.json` in the build ZIP rather than having it autogenerate?

**Comment by hakusaro** [comment]
> Has any thought been given to shipping a config.json in the build ZIP rather than having it autogenerate?

The server will persist the config file back to the disk when it needs to persist new changes. This means that if we add new items, it will always be updated with current state of the world.

This is an intentional design decision. For the vast majority of users, this means that the config file is always updated with the latest changes and is accurate. Users familiar with *nix are often accustomed to managing lots of config files in a *nix like way, amounting to a pile of different files with different rights and choices. I disagree with this philosophy: we want to be as user friendly as possible.

> Has any thought been given to shipping a config.json in the build ZIP rather than having it autogenerate?

Doing this would complicate the process of installing TShock, and I don't see that changing. Imagine we ship a `config.json.example` -- then a user has to copy it and then they never get updates from us ever again and have to always learn that we changed the config file and manually port over changes. Imagine we just ship a `config.json` -- then a user must take care to not overwrite their own. Right now, the installation process is just "extract and run the exe" and update is "extract, copy all files over, run exe." This is simple.

> How is the current config.json sorted?

It's probably "sorted" by the order of the mapped class that's serialized to the disk. We tend to group related things together. It attempts to be grouped by logical function.

If you like sorting it alphabetically for the purposes of editing, you can do that quite easily when you load the file in an editor.

**Reply by kubedzero** [reply]
> config file is always updated with the latest changes and is accurate

Makes sense. In-place upgrades are more straightforward as well, as opposed to *nix package upgrades that have complicated merge/keep/replace strategies when configs change. Thanks for sharing. 

> you can do that quite easily when you load the file in an editor

Yes, but this is a temporary edit. The second I run `/reload` or restart the server, it goes back to the random order as the serialization/deserialization happens. An unfortunate side effect of keeping the config always up-to-date I suppose. 

Digging in a little bit, I see this [ConfigFile.cs](https://github.com/Pryaxis/TShock/blob/ad82afd8001a585bcb2045900b082aea9d6e3455/TShockAPI/ConfigFile.cs) class seems to have a 1:1 ordered mapping between the object declaration and the JSON output's order. Would any consideration be given to a pull request that reorders these items to be more logical? For example, grouping port/password/login settings, grouping save-related settings, world states (creep, xmas, etc), and so on?

**Reply by hakusaro** [reply]
> Would any consideration be given to a pull request that reorders these items to be more logical? For example, grouping port/password/login settings, grouping save-related settings, world states (creep, xmas, etc), and so on?

Yes! For sure!

> Yes, but this is a temporary edit.

Yeah -- definitely have to resort. We just don't think that alphabetical is the best choice.

**Reply by kubedzero** [reply]
> Yes! For sure!

Cool! I'll see if I can put something together.

> don't think that alphabetical is the best choice

Agreed, or at least not without package-style naming (`save.autosave.interval`, `save.auotsave.enabled`, `server.port`, `server.password` etc.)

**Comment by kubedzero** [comment]
Merged in https://github.com/Pryaxis/TShock/pull/2020

Thanks all for reviewing!
